### PR TITLE
[Glimmer2] Migrating hash helper

### PR DESCRIPTION
**Pull request by chadhietala**
This migrate the hash helper to glimmer2 and moves tests over as well. As a result I remove a contextual-component flag that was still in the hash tests. This also adds tests for `yield`ing a hash which was not covered at all :smile: /cc @chancancode

### CODE REVIEW

File: packages/ember-glimmer/tests/integration/helpers/hash-test.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    });
+    this.assertText('Marisa Arbeo');
+
+    this.runTask(() => this.rerender());
+
+    this.assertText('Marisa Arbeo');
+
+    this.runTask(() => set(this.context, 'firstName', 'Sergio'));
+
+    this.assertText('Sergio Arbeo');
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/helpers/hash-test.js]: Should test resetting to `Marisa` here

File: packages/ember-glimmer/tests/integration/helpers/hash-test.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    this.runTask(() => set(this.context, 'lastName', 'Smith'));
+
+    this.assertText('Sergio Smith');
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/helpers/hash-test.js]: Should test resetting to `Marisa Arbeo` here

Comment (chadhietala) [inline packages/ember-glimmer/tests/integration/helpers/hash-test.js]: Fixed

File: packages/ember-glimmer/tests/integration/helpers/hash-test.js (hunk 1 above)
Comment (chadhietala) [inline packages/ember-glimmer/tests/integration/helpers/hash-test.js]: Fixed

**Comment by chancancode** [comment]
I think we probably want to leave the htmlbars implementation alone (i.e. duplicating the code, instead of symlinking). The final implementation of the glimmer version will probably be a bit different.

(I think we ultimately want to implement it as an "internal" helper like inline if/unless, where the helper gets the `EvaluatedArgs` instead of the values, so we can avoid re-ifying the keys that are not actually used.)

File: packages/ember-glimmer/tests/integration/helpers/hash-test.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    this.runTask(() => set(this.context, 'firstName', 'Chad'));
+
+    this.assertText('Chad');
+
+    this.runTask(() => set(this.context, 'firstName', 'Balint'));
+
+    this.assertText('Balint');
+  }
+});
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/helpers/hash-test.js]: New line between L89/90 to balance L6

File: packages/ember-glimmer/lib/helpers/hash.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+   Use the `{{hash}}` helper to create a hash to pass as an option to your
+   components. This is specially useful for contextual components where you can
+   just yield a hash:
+
+   ```handlebars
+   {{yield (hash
+      name='Sarah'
+      title=office
+   )}}
+   ```
```
Comment (chancancode) [inline packages/ember-glimmer/lib/helpers/hash.js]: It's a bit ironic that we don't actually have a test for this case :kissing:

Can you quickly add a test for yielding a hash? I *think* it won't work for glimmer yet, so it probably needs to be `@htmlbars`'ed

**Comment by chancancode** [comment]
Other than those line comments and the symlink issue, looks great to me! (We don't have to do the internal helper optimization now, unless you want to.)

File: packages/ember-glimmer/lib/helpers/hash.js (hunk 4 above)
Comment (chadhietala) [inline packages/ember-glimmer/lib/helpers/hash.js]: I added these and they seem to Just Work™. I wrote tests for internal, external, and mixed hash yields.

File: packages/ember-htmlbars/tests/helpers/hash_test.js [hunk 5]
```diff
@@ -7,67 +7,65 @@ import { runAppend, runDestroy } from 'ember-runtime/tests/utils';
 
 var view;
 
-if (isEnabled('ember-contextual-components')) {
```
Comment (chadhietala) [inline packages/ember-htmlbars/tests/helpers/hash_test.js]: This was introduced in 2.3.0. The rest of changes in this file are just formatting.

**Comment by chancancode** [comment]
@chadhietala sorry :/ I was only saying that we should keep the implementation separate, what you did with the tests before was just fine

File: packages/ember-htmlbars/lib/helpers/hash.js [hunk 6]
```diff
@@ -32,4 +27,4 @@
 
 export default function hashHelper(params, hash, options) {
   return hash;
-}
+}
```
Comment (chancancode) [inline packages/ember-htmlbars/lib/helpers/hash.js]: ProTip™: https://github.com/SamVerschueren/vscode-final-newline

**Comment by chadhietala** [comment]
@chancancode Won't we loose the test coverage there for some period time then? Maybe I don't understand.

File: packages/ember-glimmer/lib/helpers/hash.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+   @param {Object} options
+   @return {Object} Hash
+   @public
+ */
+
+function hash(params, hash, options) {
+  return hash;
+}
+
+export default helper(hash);
```
Comment (chancancode) [inline packages/ember-glimmer/lib/helpers/hash.js]: https://github.com/SamVerschueren/vscode-final-newline

File: packages/ember-glimmer/tests/integration/helpers/hash-test.js [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+    this.assertText('Chad');
+
+    this.runTask(() => set(this.context, 'firstName', 'Balint'));
+
+    this.assertText('Balint');
+  }
+
+  ['@test should yield hash of external properties']() {
+
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/helpers/hash-test.js]: :scissors:

File: packages/ember-glimmer/tests/integration/helpers/hash-test.js [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+    this.registerComponent('foo-bar', {
+      template: `{{yield (hash firstName=attrs.name)}}`
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/helpers/hash-test.js]: :scissors: `attrs.`

File: packages/ember-glimmer/tests/integration/helpers/hash-test.js [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+      init() {
+        this._super();
+        fooBarInstance = this;
+        this.firstName = 'Chad';
+      }
+    });
+
+    this.registerComponent('foo-bar', {
+      ComponentClass: FooBarComponent,
+      template: `{{yield (hash firstName=firstName lastName=attrs.lastName)}}`
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/helpers/hash-test.js]: :scissors: `attrs.`

File: packages/ember-glimmer/lib/helpers/hash.js (hunk 7 above)
Comment (chadhietala) [inline packages/ember-glimmer/lib/helpers/hash.js]: There is a line here. Probably a github issue.

**Comment by chancancode** [comment]
Merged in 0e8e343

**Pull request closed without merging**